Mathlib/Data/Real/IsNonarchimedean.lean

```diff
@@ -0,0 +1,138 @@
+    intros x hx y hy
+    rcases le_total ‖g x‖ ‖g y‖ with h|h
+    · exact ⟨y, hy, by simp [h]⟩
+    · exact ⟨x, hx, by simp [h]⟩
+  obtain ⟨x, hx, hx'⟩ := Finset.sup'_mem (α := ℝ) ((‖g ·‖) '' t) (by simpa [sup_eq_max] using ht') t
+    ht (‖g ·‖) (fun _ ↦ Set.mem_image_of_mem _)
+  refine ⟨x, fun _ ↦ by simpa using hx, ?_⟩
+  dsimp only at hx'
+  rw [hx']
+  exact ht.norm_sum_le_sup'_norm _
+
+/-- Given a nonarchimedean additive group seminorm `f` on `α`, a function `g : β → α` and a
+  nonempty finset `t : Finset β`, we can always find `b : β` belonging to `t` such that
+  `f (t.sum g) ≤ f (g b)` . -/
+theorem finset_image_add_of_nonempty {F α β : Type*} [AddCommGroup α] [FunLike F α ℝ]
+    [AddGroupSeminormClass F α ℝ] [Nonempty β] {f : F} (hna : IsNonarchimedean f)
+    (g : β → α) {t : Finset β} (ht : t.Nonempty) :
+    ∃ b : β, (b ∈ t) ∧ f (t.sum g) ≤ f (g b) := by
+  obtain ⟨b, hbt, hbf⟩ := finset_image_add hna g t
+  exact ⟨b, hbt ht, hbf⟩
+
+/-- Given a nonarchimedean additive group seminorm `f` on `α`, a function `g : β → α` and a
+  multiset `s : Multiset β`, we can always find `b : β`, belonging to `s` if `s` is nonempty,
+  such that `f (t.sum g) ≤ f (g b)` . -/
+theorem multiset_image_add {F α β : Type*} [AddCommGroup α] [FunLike F α ℝ]
+    [AddGroupSeminormClass F α ℝ] [Nonempty β] {f : F} (hna : IsNonarchimedean f)
+    (g : β → α) (s : Multiset β) :
+    ∃ b : β, (s ≠ 0 → b ∈ s) ∧ f (Multiset.map g s).sum ≤ f (g b) := by
+  inhabit β
+  induction s using Multiset.induction_on with
+  | empty => simp
+  | @cons a t hM =>
+      obtain ⟨M, hMs, hM⟩ := hM
+      by_cases hMa : f (g M) ≤ f (g a)
+      · refine ⟨a, by simp, ?_⟩
+        · rw [Multiset.map_cons, Multiset.sum_cons]
+          exact le_trans (hna _ _) (max_le (le_refl _) (le_trans hM hMa))
+      · rw [not_le] at hMa
+        rcases eq_or_ne t 0 with rfl|ht
+        · exact ⟨a, by simp, by simp⟩
+        · refine ⟨M, ?_, ?_⟩
+          · simp [hMs ht]
+          rw [Multiset.map_cons, Multiset.sum_cons]
+          exact le_trans (hna _ _) (max_le hMa.le hM)
+
+/-- Given a nonarchimedean additive group seminorm `f` on `α`, a function `g : β → α` and a
+  nonempty multiset `s : Multiset β`, we can always find `b : β` belonging to `s` such that
+  `f (t.sum g) ≤ f (g b)` . -/
+theorem multiset_image_add_of_nonempty {F α β : Type*} [AddCommGroup α] [FunLike F α ℝ]
+    [AddGroupSeminormClass F α ℝ] [Nonempty β] {f : F} (hna : IsNonarchimedean f)
+    (g : β → α) {s : Multiset β} (hs : s ≠ 0) :
+    ∃ b : β, (b ∈ s) ∧ f (Multiset.map g s).sum ≤ f (g b) := by
+  obtain ⟨b, hbs, hbf⟩ := multiset_image_add hna g s
+  exact ⟨b, hbs hs, hbf⟩
+
+/-- If `f` is a nonarchimedean additive group seminorm on a commutative ring `α`, `n : ℕ`, and
+  `a b : α`, then we can find `m : ℕ` such that `m ≤ n` and
+  `f ((a + b) ^ n) ≤ (f (a ^ m)) * (f (b ^ (n - m)))`. -/
+theorem add_pow_le {F α : Type*} [CommRing α] [FunLike F α ℝ]
```

Comment:
And again: if this lemma is useful we should formulate it for `[SeminormedRing R] [IsUltrametricDist R]` and refer to that here.